[libra framework] First cut at specifying DesignatedDealer

shazqadeer opened the pull request:
## Motivation

This PR adds specifications for DesignatedDealer module.  The verification is currently turned off due to performance problems in Move Prover.  I will revisit the verification in a bit.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Existing tests and pending Move Prover verification

## Related PRs

moeziniaf commented on language/stdlib/modules/DesignatedDealer.move:
```diff
@@ -62,69 +64,81 @@ module DesignatedDealer {
+                assert(new_upperbound < *Vector::borrow(tiers, next_tier_index), EINVALID_TIER_START);
+            };
+        };
+        if (*tier > new_upperbound && tier_index > 0) {
+            let prev_tier_index = tier_index - 1;
+            assert(new_upperbound > *Vector::borrow(tiers, prev_tier_index), EINVALID_TIER_START);
```
thanks 👍

moeziniaf commented on language/stdlib/modules/DesignatedDealer.move:
```diff
@@ -62,69 +64,81 @@ module DesignatedDealer {
+        let cur_inflow = dealer.window_inflow;
+        let new_inflow = cur_inflow + amount;
         let tiers = &mut dealer.tiers;
         // If the tier_index is one past the bounded tiers, minting is unbounded
```
I think we should remove this comment, since line 119 removal (correctly) removes this functionality

moeziniaf commented on language/stdlib/modules/DesignatedDealer.move:
```diff
@@ -62,69 +64,81 @@ module DesignatedDealer {
-        if (tier_index == number_of_tiers) {
-            *tier_check = true;
-        } else {
+        assert(tier_index <= number_of_tiers, EINVALID_TIER_INDEX);
```
I think this should be a strict inequality
```suggestion
        assert(tier_index < number_of_tiers, EINVALID_TIER_INDEX);
```

moeziniaf commented on language/stdlib/modules/DesignatedDealer.move:
```diff
@@ -62,69 +64,81 @@ module DesignatedDealer {
-            *tier_check = (cur_inflow + amount <= tier_upperbound);
+            assert(new_inflow <= tier_upperbound, EINVALID_AMOUNT_FOR_TIER);
         };
-        if (*tier_check) {
-            dealer.window_inflow = cur_inflow + amount;
-        };
-        *tier_check
```
💯

moeziniaf commented on language/stdlib/modules/DesignatedDealer.move:
```diff
@@ -148,19 +161,36 @@ module DesignatedDealer {
+            assert forall x: address where exists<Dealer>(x):
+                         forall i: u64, j: u64 where 0 <= i && i < j && j < len(global<Dealer>(x).tiers):
+                            global<Dealer>(x).tiers[i] < global<Dealer>(x).tiers[j];
+            assume false; // TODO: It ought to be possible to remove this once the memory splitting across modules lands.
```
what is the meaning of this line?

moeziniaf commented on language/stdlib/modules/DesignatedDealer.move:
```diff
@@ -148,19 +161,36 @@ module DesignatedDealer {
+        // modifies global<Dealer>@dd_addr.{window_start, window_inflow, mint_event_handle}
+        ensures {let dealer = global<Dealer>(dd_addr);
+                {let current_time = LibraTimestamp::spec_now_microseconds();
+                    (dealer.window_start == current_time && dealer.window_inflow == amount) ||
+                    (dealer.window_start + spec_window_length() > current_time && dealer.window_inflow == old(dealer.window_inflow) + amount)
```
👍

moeziniaf commented on language/stdlib/modules/DesignatedDealer.move:
```diff
@@ -148,19 +161,36 @@ module DesignatedDealer {
+                    (dealer.window_start + spec_window_length() > current_time && dealer.window_inflow == old(dealer.window_inflow) + amount)
+                }};
+        ensures tier_index <= len(old(global<Dealer>(dd_addr)).tiers);
+        ensures tier_index < len(old(global<Dealer>(dd_addr)).tiers) ==> global<Dealer>(dd_addr).window_inflow <= old(global<Dealer>(dd_addr)).tiers[tier_index];
```
is ==> == iff? or implies a statement?

moeziniaf commented:
FYI https://github.com/libra/libra/pull/4978

shazqadeer commented:
@moeziniaf and I discussed offline.  I was not aware that the feature of unbounded minting is now deprecated.  I have removed this feature from the PR and updated the spec I wrote.

shazqadeer commented:
@bors-libra r=moeziniaf

bors-libra commented:
:pushpin: Commit 154bef2 has been approved by `moeziniaf`

<!-- @bors-libra r=moeziniaf 154bef22d3e4186729de8321694d0cff1737eb0a -->

bors-libra commented:
:hourglass: Testing commit 154bef22d3e4186729de8321694d0cff1737eb0a with merge d8dbbd1fb54980314d3dd25b6cdb021045bd7bc2...

wrwg commented:
> [move-prover] Move Prover test (`language/move-prover/cargo test`) on commit [154bef2](https://github.com/libra/libra/commit/154bef22d3e4186729de8321694d0cff1737eb0a) did **fail**. See [details](https://github.com/mimoo/mirai-bot/runs/856463342?check_suite_focus=true#step:7:1).

Can you please be sure that MIRAI bot passes before you merge so we add not any additional failures to the already failing things? Libra.move fails at head because yesterday some changes were submitted which break it. You can add `verify = false` to the module to disable it for now.

bors-libra commented:
:sunny: Test successful - [checks-actions_land_blocking_test](https://github.com/libra/libra/runs/856478916), [checks-circle_commit_workflow](https://circleci.com/workflow-run/2fcf47b7-0337-4824-86eb-5e65a90a9478?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link)
Approved by: moeziniaf
Pushing d8dbbd1fb54980314d3dd25b6cdb021045bd7bc2 to master...
